### PR TITLE
home maintenance error handling

### PR DESCRIPTION
**Pull request by rhit-daviscl**
## Jira Ticket:
RBH 65 Home Maintenance Error Handling
## Description of Work
In this PR, I added code to the modals of the Home Maintenance to stop user errors on those modals.
## Acceptance Criteria

- User cannot cause issues in Modals

## How to Test

- Manual test by trying to do things the modal shouldn't cover.

### List of Changes:

- Invitations stop the user from inviting someone twice, inviting people in the house already, inputting a non-email, and from inviting themselves.
- Edit permissions does not allow a User to change their own permissions, just other peoples.
- Remove Roommate will not allow a person to be removed if they are the only owner in a house
- Leave house will not allow someone to leave if they are the only user with owner role.
- Whenever a person leaves a house or deletes one, all permissions tied to those people are removed.

### Screenshots

| Image | Description |
| ----- | ----------- |
| image|User tried inviting someone twice|
|image|User tried to remove the only Owner|
|image|User tried changing their own permissions|

### CODE REVIEW

File: app/src/pages/homes.tsx [hunk 1]
```diff
@@ -98,30 +111,43 @@ const HomesPage: NextPage = () => {
...
+            setEditPermissionsModalOpen(false);
+        }
     }
 
     const handleInvite = async (event: React.FormEvent<HTMLFormElement>) => {
         event.preventDefault();
         const form = (event.target as HTMLFormElement);
         if(!selectedHome) return;
-        await inviteRoommate.mutateAsync({
+        if (!/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(form.email.value)) {
```
Comment (RealJasomo) [inline app/src/pages/homes.tsx:137]: I would use zod to validate this or their regex, they use a more extensive one to handle odd cases of emails

File: app/src/server/db/HomeService.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import { Permission } from "@aws-sdk/client-s3";
```
Comment (RealJasomo) [inline app/src/server/db/HomeService.ts:1]: Was this an accidental import?

**Review by RealJasomo — CHANGES_REQUESTED** [review]

File: app/src/server/db/HomeService.ts (hunk 2 above)
Comment (rhit-daviscl) [reply app/src/server/db/HomeService.ts]: Yes it was. I was using the Permission import incorrectly. I removed it.

File: app/src/pages/homes.tsx (hunk 1 above)
Comment (rhit-daviscl) [reply app/src/pages/homes.tsx]: I added this check in the latest commit.

File: app/src/server/db/HomeService.ts [hunk 3]
```diff
@@ -10,4 +10,43 @@ export const canUserViewHome = async (id: string, homeId: string, db: PrismaClie
...
+        where: {
+            homeId: homeId,
+            email: email
+        }
+    }))
+}
+
+export const userIsInHome = async (userId: string, homeId: string, db: PrismaClient) => {
+    return !!(await db.occupies.findFirst({
+        select: {
```
Comment (RealJasomo) [inline app/src/server/db/HomeService.ts:29]: This is exactly the same as userCanViewHome function at top

**Review by RealJasomo — CHANGES_REQUESTED** [review]
Remove duplicated function

**Review by RealJasomo — APPROVED** [review]
Thanks for the change, looks good now

**Review by RealJasomo — APPROVED** [review]
Now looks good thanks

**Pull request merged**